sklearn/model_selection/_validation.py

```diff
@@ -1167,7 +1186,13 @@ def learning_curve(estimator, X, y, groups=None,
         train_test_proportions = []
         for train, test in cv_iter:
             for n_train_samples in train_sizes_abs:
-                train_test_proportions.append((train[:n_train_samples], test))
+                if stratify and n_train_samples < n_max_training_samples:
+                    cur_train, useless_test = train_test_split(
+                        train, stratify=y[train], train_size=n_train_samples,
```

Comment:
For the binary case, something like:

```
from sklearn.model_selection._split import _approximate_mode

def _stratified_random_sampling(y_train, n_samples_per_class):
    class0 = np.where(y_train == 0)[0]
    class1 = np.where(y_train == 1)[0]
    train_inner0 = np.random.choice(class0, n_samples_per_class[0], replace=False)
    train_inner1 = np.random.choice(class1, n_samples_per_class[1], replace=False)
    return np.hstack((train_inner0, train_inner1))
```
and in the for loop:

```
n_samples_per_class = _approximate_mode(np.bincount(y_train), n_train_samples, 0)
train_inner = _stratified_random_sampling(y[train], n_samples_per_class)
              
train_test_proportions.append((train[train_inner], test))
```

Reply:
Thanks for revew @imad3v! Good catch. I see the following options here:

1. Keep at is as, just add another check 
`n_train_samples == n_max_training_samples - num_classes`
which would trigger a warning, so that the user would understand the problem and could change dataset or value of train_sizes parameter (say, reduce max train size from 1.0 to 0.9).

2. Add the check and the warning from the previous item, but in this case use the whole set as a training (i.e. add these 1 or 2 instances, which are not sufficient for useless test set).

3. Add new method to `_split` module with 25 lines of copy-pasted code from `train_test_split()` and `StratifiedShuffleSplit._iter_indices()`. Your code snippet doesn't cover multiclass/multilabel case and other stuff that is implemented in the aforementioned method.

4. Refactor `StratifiedShuffleSplit._iter_indices()` so that it would consist of 2 calls to 2 method and 1 validation check between these 2 calls:
```python
    def _iter_indices(self, X, y, groups=None):
        class_counts, n_classes, n_test, n_train, y_indices = self.estimate_counts(X, y)
        if np.min(class_counts) < 2:
            raise ValueError("The least populated class in y has only 1"
		//(and other 2 checks)
        yield from self.stratified_split(class_counts, n_classes, n_test, n_train, y_indices)
```
then call these methods without checks. Basically, the only reason why we can't reuse this method is an excessive check for our use-case.

All options seem to be far from ideal, but I'm inclined to the last one. What do you think?

Reply:
I wouldn't go for the first 2 options as they avoid the problem instead of solving it. Option 4 looks good.

Another option is to modify `StratifiedShuffleSplit` to accept `test_size=0`, and avoid the validation check in this case. I don't know if this change would break other features or not.

Reply:
@imad3v please, review the changes.
Also could you please take a look at the failed build - seems like the reason is not my changes.

Reply:
This works, thanks!